src/lib/fhirValidator.js

```
@@ -0,0 +1,86 @@
+import Ajv from 'ajv';
+import request from 'superagent-bluebird-promise';
+
+import ValidationError from './errors';
+import config from '../config';
+
+const ajv = new Ajv({ extendRefs: true });
+
+class Fhir {
+	constructor() {
+		this.getConformance();
+	}
+	getConformance() {
+		if (this.conformance !== undefined) {
+			return Promise.resolve(this.conformance);
+		}
+
+		return request.get(config.fhirSchemaUrl)
+			.then((res) => {
+				const schema = JSON.parse(res.text);
+				Object.keys(schema.resources).forEach((key) => {
+					schema.resources[key].types = Object.assign({},
+						schema.resources[key].types, schema.types);
+				});
+				this.conformance = {
+					types: schema.types,
+					resources: schema.resources,
+				};
+				return this.conformance;
+			});
+	}
+
+	isValidResourceType(resourceType) {
+		if (resourceType.includes('.')) return new Promise((resolve, reject) => reject(false));
+		return this.getConformance()
+			.then(conform => Object.prototype.hasOwnProperty.call(conform.resources, resourceType));
+	}
+
+	validate(data) {
+		let result;
+		let	resource;
+		let	valid;
+
+		result = {
+			resource: null,
+			errors: [],
+		};
+
+		// Ensure that the data provided is a javascript object.
+		// If it is a string then attempt to parse it
+		try {
+			if (typeof data === 'string') resource = JSON.parse(data);
+			else if (typeof data === 'object') resource = data;
+			else {
+				result.errors.push(new ValidationError(
+					'Data provided must be a string or plain javascript object',
+				));
+				return Promise.reject(result);
```

throw new ValidationError() or return Promise.reject(new ValidationError()) instead of returning result